ui/src/app/workflows/components/workflow-dag/workflow-dag.tsx

```diff
@@ -48,94 +54,99 @@ function nodeLabel(n: NodeStatus) {
-    return v;
-})();
-
-export class WorkflowDag extends React.Component<WorkflowDagProps, WorkflowDagRenderOptions> {
-    private graph: Graph;
-
-    constructor(props: Readonly<WorkflowDagProps>) {
-        super(props);
-        this.state = {
-            expandNodes: new Set(),
-            showArtifacts: localStorage.getItem('showArtifacts') !== 'false'
-        };
-    }
+const classNames = {
+    Artifact: true,
+    Suspended: true,
+    Collapsed: true,
+    ...Object.entries(NODE_PHASE).reduce((acc, [, value]) => ({...acc, [value]: true}), {})
+};
 
-    private get artifactRepository() {
-        return this.props.artifactRepositoryRef?.artifactRepository || {};
-    }
+export function WorkflowDag(props: WorkflowDagProps) {
+    let graph: Graph;
+    const artifactRepository = props.artifactRepositoryRef?.artifactRepository || {};
+    const tags: Record<string, boolean> = Object.values(props.nodes || {}).reduce((acc, node) => ({...acc, [getNodeLabelTemplateName(node)]: true}), {});
 
-    public render() {
-        this.prepareGraph();
-
-        const tags: {[key: string]: boolean} = {};
-        Object.values(this.props.nodes || {}).forEach(n => (tags[getNodeLabelTemplateName(n)] = true));
-
-        return (
-            <GraphPanel
-                storageScope='workflow-dag'
-                graph={this.graph}
-                nodeGenresTitle={'Node Type'}
-                nodeGenres={genres}
-                nodeClassNamesTitle={'Node Phase'}
-                nodeClassNames={classNames}
-                nodeTagsTitle={'Template'}
-                nodeTags={tags}
-                nodeSize={this.props.nodeSize || 32}
-                defaultIconShape='circle'
-                hideNodeTypes={true}
-                hideOptions={this.props.hideOptions}
-                selectedNode={this.props.selectedNodeId}
-                onNodeSelect={id => this.selectNode(id)}
-                options={<WorkflowDagRenderOptionsPanel {...this.state} onChange={workflowDagRenderOptions => this.saveOptions(workflowDagRenderOptions)} />}
-            />
-        );
-    }
+    const [expandNodes, setExpandNodes] = useState(new Set(''));
+    const [showArtifacts, setShowArtifacts] = useState(localStorage.getItem('showArtifacts') !== 'false');
 
-    private saveOptions(newChanges: WorkflowDagRenderOptions) {
-        localStorage.setItem('showArtifacts', newChanges.showArtifacts ? 'true' : 'false');
-        this.setState(newChanges);
+    prepareGraph();
+
+    useEffect(() => {
+        prepareGraph();
+    }, [props.nodes, expandNodes, showArtifacts]);
+
+    function getNode(nodeId: string) {
+        return props.nodes[nodeId];
     }
 
-    private getNode(nodeId: string): NodeStatus {
-        const node: NodeStatus = this.props.nodes[nodeId];
-        if (!node) {
-            return null;
+    function prepareGraph() {
+        const {edges, nodes} = (graph = new Graph());
+
+        if (!props.nodes) {
+            return;
         }
```

Comment:
was this not needed? the check does still exist in a few places

Reply:
oh this comment ended up a bit weird -- I was commenting on the `if (!node)` check that existed in the `getNode` function

Reply:
oh hold up, it returns `undefined` if it doesn't exist, so yea, this isn't necessary. good simplification! there were a few of these that took a bit for me to realize they're equivalent.

so nvm on this whole comment thread 😄